### PR TITLE
Allow trailing commas (fix #427)

### LINKED ISSUE #427: Parser for 2.12.8 should recognize trailing comma in function application
```scala
object C {
    def foo(x: Int = 0) = 0
    foo(x = 1,
    )
}
```
This actually parses in `scalac` `2.12.8` despite the `,`. Note that it will fail to parse if the `)` is on the same line as the `,`!

```
C.scala:4: error: illegal start of simple expression
    )
    ^
rsc.util.CrashException: illegal start of simple expression
	at rsc.util.CrashUtil.crash(CrashUtil.scala:10)
	at rsc.util.CrashUtil.crash$(CrashUtil.scala:9)
	at rsc.util.package$.crash(package.scala:5)
	at rsc.parse.scala.Messages.reportOffset(Messages.scala:39)
	at rsc.parse.scala.Messages.reportOffset$(Messages.scala:34)
	at rsc.parse.scala.Parser.reportOffset(Parser.scala:13)
	at rsc.parse.scala.Terms.simpleTerm(Terms.scala:320)
	at rsc.parse.scala.Terms.prefixTerm(Terms.scala:239)
	at rsc.parse.scala.Terms.postfixTerm(Terms.scala:198)
	at rsc.parse.scala.Terms.postfixTerm$(Terms.scala:189)
	at rsc.parse.scala.Parser.postfixTerm(Parser.scala:13)
	at rsc.parse.scala.Terms.term1(Terms.scala:142)
	at rsc.parse.scala.Terms.term1$(Terms.scala:50)
	at rsc.parse.scala.Parser.term1(Parser.scala:13)
	at rsc.parse.scala.Terms.$anonfun$term$1(Terms.scala:40)
	at rsc.parse.scala.Wildcards.wrapEscapingTermWildcards(Wildcards.scala:56)
	at rsc.parse.scala.Wildcards.wrapEscapingTermWildcards$(Wildcards.scala:53)
	at rsc.parse.scala.Parser.wrapEscapingTermWildcards(Parser.scala:13)
	at rsc.parse.scala.Terms.term(Terms.scala:38)
	at rsc.parse.scala.Terms.term$(Terms.scala:34)
	at rsc.parse.scala.Parser.term(Parser.scala:13)
	at rsc.parse.scala.Terms.$anonfun$termArgs$2(Terms.scala:29)
	at rsc.parse.scala.Helpers.tokenSeparated(Helpers.scala:103)
	at rsc.parse.scala.Helpers.commaSeparated(Helpers.scala:58)
	at rsc.parse.scala.Helpers.commaSeparated$(Helpers.scala:57)
	at rsc.parse.scala.Parser.commaSeparated(Parser.scala:13)
	at rsc.parse.scala.Terms.$anonfun$termArgs$1(Terms.scala:29)
	at rsc.parse.scala.Helpers.inParens(Helpers.scala:81)
	at rsc.parse.scala.Helpers.inParens$(Helpers.scala:79)
	at rsc.parse.scala.Parser.inParens(Parser.scala:13)
	at rsc.parse.scala.Terms.termArgs(Terms.scala:28)
	at rsc.parse.scala.Terms.termArgs$(Terms.scala:22)
	at rsc.parse.scala.Parser.termArgs(Parser.scala:13)
	at rsc.parse.scala.Terms.simpleTermRest(Terms.scala:343)
	at rsc.parse.scala.Terms.simpleTerm(Terms.scala:324)
	at rsc.parse.scala.Terms.prefixTerm(Terms.scala:239)
	at rsc.parse.scala.Terms.postfixTerm(Terms.scala:198)
	at rsc.parse.scala.Terms.postfixTerm$(Terms.scala:189)
	at rsc.parse.scala.Parser.postfixTerm(Parser.scala:13)
	at rsc.parse.scala.Terms.term1(Terms.scala:142)
	at rsc.parse.scala.Terms.term1$(Terms.scala:50)
	at rsc.parse.scala.Parser.term1(Parser.scala:13)
	at rsc.parse.scala.Templates.$anonfun$templateBraces$2(Templates.scala:134)
	at rsc.parse.scala.Helpers.inBraces(Helpers.scala:67)
	at rsc.parse.scala.Helpers.inBraces$(Helpers.scala:65)
	at rsc.parse.scala.Parser.inBraces(Parser.scala:13)
	at rsc.parse.scala.Templates.$anonfun$templateBraces$1(Templates.scala:92)
	at rsc.parse.scala.Wildcards.banEscapingWildcards(Wildcards.scala:18)
	at rsc.parse.scala.Wildcards.banEscapingWildcards$(Wildcards.scala:15)
	at rsc.parse.scala.Parser.banEscapingWildcards(Parser.scala:13)
	at rsc.parse.scala.Templates.templateBraces(Templates.scala:89)
	at rsc.parse.scala.Templates.defnTemplate(Templates.scala:45)
	at rsc.parse.scala.Templates.defnTemplate$(Templates.scala:22)
	at rsc.parse.scala.Parser.defnTemplate(Parser.scala:13)
	at rsc.parse.scala.Defns.defnObject(Defns.scala:84)
	at rsc.parse.scala.Defns.defnObject$(Defns.scala:81)
	at rsc.parse.scala.Parser.defnObject(Parser.scala:13)
	at rsc.parse.scala.Sources.$anonfun$packageStats$1(Sources.scala:77)
	at rsc.parse.scala.Wildcards.banEscapingWildcards(Wildcards.scala:18)
	at rsc.parse.scala.Wildcards.banEscapingWildcards$(Wildcards.scala:15)
	at rsc.parse.scala.Parser.banEscapingWildcards(Parser.scala:13)
	at rsc.parse.scala.Sources.packageStats(Sources.scala:53)
	at rsc.parse.scala.Sources.$anonfun$sourceStats$1(Sources.scala:48)
	at rsc.parse.scala.Wildcards.banEscapingWildcards(Wildcards.scala:18)
	at rsc.parse.scala.Wildcards.banEscapingWildcards$(Wildcards.scala:15)
	at rsc.parse.scala.Parser.banEscapingWildcards(Parser.scala:13)
	at rsc.parse.scala.Sources.sourceStats(Sources.scala:18)
	at rsc.parse.scala.Sources.source(Sources.scala:15)
	at rsc.parse.scala.Sources.source$(Sources.scala:13)
	at rsc.parse.scala.Parser.source(Parser.scala:13)
	at rsc.parse.scala.Parser.parse(Parser.scala:42)
	at rsc.Compiler.$anonfun$parse$2(Compiler.scala:89)
	at scala.collection.immutable.List.flatMap(List.scala:338)
	at rsc.Compiler.parse(Compiler.scala:78)
	at rsc.Compiler.$anonfun$phases$1(Compiler.scala:68)
	at rsc.Compiler.$anonfun$run$2(Compiler.scala:36)
	at rsc.Compiler.$anonfun$run$2$adapted(Compiler.scala:33)
	at scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:792)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:791)
	at rsc.Compiler.run(Compiler.scala:33)
	at rsc.cli.Main$.process(Main.scala:35)
	at rsc.cli.Main$.main(Main.scala:14)
	at rsc.cli.Main.main(Main.scala)
```

### PR DESCRIPTION
**Pull request by ShaneDelmore**

### CODE REVIEW

File: rsc/src/main/scala/rsc/parse/scala/Scanners.scala [hunk 1]
```diff
@@ -166,4 +166,22 @@ trait Scanners {
       in.nextToken()
     }
   }
+
+  def optTrailingComma(): Unit = {
```
Comment (wiwa) [inline rsc/src/main/scala/rsc/parse/scala/Scanners.scala]: Do we really need to use `restore(snapshot)`?

File: examples/syntactic/src/main/scala/427.scala [hunk 2]
```diff
@@ -0,0 +1,5 @@
+object C {
```
Comment (wiwa) [inline examples/syntactic/src/main/scala/427.scala]: We also want an error case, like `foo(x = 1,)`, and test with comments too.

File: rsc/src/main/scala/rsc/parse/scala/Scanners.scala (hunk 1 above)
Comment (ShaneDelmore) [reply rsc/src/main/scala/rsc/parse/scala/Scanners.scala]: If we don't use restore(snapshot) then the scanner is out of sync with the in object.

**Comment by ShaneDelmore** [comment]
Benchmark after changes:
[info] Benchmark                   Mode   Cnt    Score   Error  Units
[info] RscParse.run              sample  2367   21.132 ± 0.191  ms/op
[info] RscParse.run:run·p0.00    sample         16.941          ms/op
[info] RscParse.run:run·p0.50    sample         20.972          ms/op
[info] RscParse.run:run·p0.90    sample         23.462          ms/op
[info] RscParse.run:run·p0.95    sample         25.382          ms/op
[info] RscParse.run:run·p0.99    sample         27.512          ms/op
[info] RscParse.run:run·p0.999   sample         52.604          ms/op
[info] RscParse.run:run·p0.9999  sample        102.236          ms/op
[info] RscParse.run:run·p1.00    sample        102.236          ms/op
[success] Total time: 112 s, completed Jul 29, 2019 3:10:13 PM

Benchmark before changes:
[info] Benchmark                   Mode   Cnt   Score   Error  Units
[info] RscParse.run              sample  2457  20.355 ± 0.130  ms/op
[info] RscParse.run:run·p0.00    sample        16.843          ms/op
[info] RscParse.run:run·p0.50    sample        20.349          ms/op
[info] RscParse.run:run·p0.90    sample        22.381          ms/op
[info] RscParse.run:run·p0.95    sample        24.642          ms/op
[info] RscParse.run:run·p0.99    sample        26.242          ms/op
[info] RscParse.run:run·p0.999   sample        34.191          ms/op
[info] RscParse.run:run·p0.9999  sample        37.356          ms/op
[info] RscParse.run:run·p1.00    sample        37.356          ms/op
[success] Total time: 106 s, completed Jul 29, 2019 3:07:57 PM

**Comment by wiwa** [comment]
Could you run benchSemanticdb and benchScalasig? To show the performance in context of a full run.

**Comment by ShaneDelmore** [comment]
master
[info] # Run complete. Total time: 00:01:42
[info] Benchmark                        Mode  Cnt    Score   Error  Units
[info] RscSemanticdb.run              sample  217  234.244 ± 5.086  ms/op
[info] RscSemanticdb.run:run·p0.00    sample       198.705          ms/op
[info] RscSemanticdb.run:run·p0.50    sample       231.735          ms/op
[info] RscSemanticdb.run:run·p0.90    sample       252.497          ms/op
[info] RscSemanticdb.run:run·p0.95    sample       268.960          ms/op
[info] RscSemanticdb.run:run·p0.99    sample       316.387          ms/op
[info] RscSemanticdb.run:run·p0.999   sample       439.878          ms/op
[info] RscSemanticdb.run:run·p0.9999  sample       439.878          ms/op
[info] RscSemanticdb.run:run·p1.00    sample       439.878          ms/op

[info] Benchmark                      Mode  Cnt    Score   Error  Units
[info] RscScalasig.run              sample  147  346.241 ± 4.964  ms/op
[info] RscScalasig.run:run·p0.00    sample       317.719          ms/op
[info] RscScalasig.run:run·p0.50    sample       343.409          ms/op
[info] RscScalasig.run:run·p0.90    sample       365.009          ms/op
[info] RscScalasig.run:run·p0.95    sample       379.899          ms/op
[info] RscScalasig.run:run·p0.99    sample       422.597          ms/op
[info] RscScalasig.run:run·p0.999   sample       429.392          ms/op
[info] RscScalasig.run:run·p0.9999  sample       429.392          ms/op
[info] RscScalasig.run:run·p1.00    sample       429.392          ms/op

Post 427 benchSemanticdb
[info] Benchmark                        Mode  Cnt    Score   Error  Units
[info] RscSemanticdb.run              sample  219  231.744 ± 4.431  ms/op
[info] RscSemanticdb.run:run·p0.00    sample       195.559          ms/op
[info] RscSemanticdb.run:run·p0.50    sample       226.755          ms/op
[info] RscSemanticdb.run:run·p0.90    sample       267.125          ms/op
[info] RscSemanticdb.run:run·p0.95    sample       271.057          ms/op
[info] RscSemanticdb.run:run·p0.99    sample       295.279          ms/op
[info] RscSemanticdb.run:run·p0.999   sample       305.136          ms/op
[info] RscSemanticdb.run:run·p0.9999  sample       305.136          ms/op
[info] RscSemanticdb.run:run·p1.00    sample       305.136          ms/op
[success] Total time: 108 s, completed Jul 29, 2019 4:20:24 PM

[info] Benchmark                      Mode  Cnt    Score   Error  Units
[info] RscScalasig.run              sample  134  378.059 ± 8.862  ms/op
[info] RscScalasig.run:run·p0.00    sample       325.583          ms/op
[info] RscScalasig.run:run·p0.50    sample       372.769          ms/op
[info] RscScalasig.run:run·p0.90    sample       422.838          ms/op
[info] RscScalasig.run:run·p0.95    sample       434.373          ms/op
[info] RscScalasig.run:run·p0.99    sample       476.630          ms/op
[info] RscScalasig.run:run·p0.999   sample       490.209          ms/op
[info] RscScalasig.run:run·p0.9999  sample       490.209          ms/op
[info] RscScalasig.run:run·p1.00    sample       490.209          ms/op
[success] Total time: 108 s, completed Jul 29, 2019 4:23:28 PM

**Pull request merged**